### PR TITLE
Set the state before saving "last" or "none" checkpoints

### PR DESCRIPTION
**Pull request by nithinraok**
Signed-off-by: nithinraok <nithinrao.koluguri@gmail.com>

## What does this PR do?

Related to issue #11451 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -643,12 +643,12 @@ def _save_last_checkpoint(self, trainer: "pl.Trainer", monitor_candidates: Dict[
             return
 
         filepath = self.format_checkpoint_name(monitor_candidates, self.CHECKPOINT_NAME_LAST)
-        trainer.save_checkpoint(filepath, self.save_weights_only)
 
         if self.last_model_path and self.last_model_path != filepath:
             trainer.strategy.remove_checkpoint(self.last_model_path)
 
         self.last_model_path = filepath
+        trainer.save_checkpoint(filepath, self.save_weights_only)
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:651]: The problem of putting this here is that if `save_checkpoint` fails (maybe the user has a bug in a `on_save_checkpoint` hook) then training stops and the previous checkpoint is gone.

Comment (nithinraok) [reply pytorch_lightning/callbacks/model_checkpoint.py:651]: Agreed, that is true but if we don't do it here, as said it's losing the previous last.ckpt state while resuming the run. 
Do you have any other suggestion?

Comment (nithinraok) [reply pytorch_lightning/callbacks/model_checkpoint.py:651]: or what might be causing this issue?

Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:651]: Followed up in the issue

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -442,6 +442,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed wrong typehint for `Trainer.lightning_optimizers` ([#11155](https://github.com/PyTorchLightning/pytorch-lightning/pull/11155))
 
 
+- Fixed bug where the path for "last" checkpoints was not getting saved correctly which caused newer runs to not remove the previous "last" checkpoint ([#11481](https://github.com/PyTorchLightning/pytorch-lightning/pull/11481))
+
+
+- Fixed bug where the path for best checkpoints was not getting saved correctly when no metric was monitor which caused newer runs to not use the best checkpoint ([#11481](https://github.com/PyTorchLightning/pytorch-lightning/pull/11481))
```
Comment (ananthsub) [inline CHANGELOG.md:448]:
```suggestion
- Fixed bug where the path for best checkpoints was not getting saved correctly when no metric was monitored which caused newer runs to not use the best checkpoint ([#11481](https://github.com/PyTorchLightning/pytorch-lightning/pull/11481))
```

**Review by ananthsub — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**